### ISSUE TITLE
Reasoning for "MUST" in resetting MINOR to zero?

### ISSUE THREAD

**Issue by nolar** [semver/semver#630]
> 8. … Patch and minor version MUST be reset to 0 when major version is incremented.

Can you please add a reason for "MUST" in the rule 8 into FAQ?

I've gone through almost all open & closed issues regarding this topic (phew! a lot!), and none contained a single explanation of "why". The closest of them all was the oldest one: #9, which only answers "how", or "what" in the best case, but not "why".

_Please note that this question is not about CI build numbers or other "implementation details", but rather about the API compatibility and risk management, i.e. human-managed versions with meaning attached._

What specifically is wrong with the following sequence of versions from the "semantic" point of view?

* `0.12.0`
* `0.13.0`
* `1.14.0`
* `1.15.0`
* `2.16.0`

Each new release adds a feature: milestones 12-16 — or, better said, feature-sets 12-16. Some of these releases break some (but not all) of the APIs/protocols/promises, and therefore increase the major version, but continue in the same milestone sequence for minor versions. All the SemVer-compatible software will handle that properly (as if versions `1.0.0-1.13.0`, `2.0.0-2.15.0` were released and instantly hidden/yanked).

It might be convincing to follow the same resetting policy for minor & patch versions for the generalisation purposes, same as the time does with hours & minutes & seconds. But these version components have different semantics: 

Semantically (i.e. "by meaning"), the patch versions (`x.y.1`, `x.y.2`) are little addons on top of the foundation that represents functionality (`x.y.0`) — as essentially defined in the SemVer's Summary.

However, the minor versions (`x.1.0`, `x.2.0`) are semantically NOT addons on top of the foundation represented by the new API (`x.0.0`). 

Unlike the "patches", the functionality is continuous (in most cases). Unless it is a full remake of the library, in which case the "functionality" is revised, and the major version upgrade really means the reset of functionality, so it makes sense to semantically start with zero in those cases (quite rare).

I'm not sure if it is worth asking to change the "MUST" word to "SHOULD" or "MAY" and let the developers decide. It was requested before, to no avail.

I also understand that to be formally compliant to SemVer, things MUST be done as written. But it would be good if the reasoning behind the rules or the spirit of the versioning schema is also explained.

Thanks for your attention.

**Comment by ljharb** [comment]
tbh i think that's a great reason - so as to PROHIBIT people from inferring additional meaning from "the minor is steadily increasing, unrelated to the major/patch". Semver conveys breakage, and should not be used to convey anything else, including "feature-sets" or "milestones".

**Comment by nolar** [comment]
Just to clarify: I didn't mean "unrelated to the patch". Quite the opposite — patches (defined as bugfixes — point 6 of the spec) are strictly bound to their relevant major.minor tuples (defined as functionality and APIs — points 7-8).

**Comment by ljharb** [comment]
Sure, thanks for clarifying - but then you're still trying to squish additional semantic meaning into a semver version, beyond what semver is supposed to solely communicate.

**Comment by nolar** [comment]
And that's the point of this issue and the trigger for it:

I inferred these meanings from the description of SemVer itself:

> _Under this scheme, version numbers and the way they change convey meaning about the underlying code and what has been modified from one version to the next._

A transition from `0.13.0` to `1.14.0` conveys the message that a few features were added, but "14" is nearly the same as "13", and we broke the API a bit. All of this is in line with the spirit of SemVer as seen by a naïve reader.

Specifically, I didn't get how it (SemVer) goes from these declarations about its purpose to the specific rules on how to reset minor versions. Which makes me assume that maybe someone in the past just put their habits or preferences into it instead of the logical and consistent values — and that knowledge is now lost.

If that meaning is not supposed to be inferred, then the SemVer introduction and description are ambiguous, or not clear about what SemVer is NOT about or NOT for. — Hence the question.

However, if you state that SemVer is about breakage only, then the minor resets should not be covered/regulated at all, leaving them to the developers, should they?

I say this not for the purpose of having a fight or question your previous decisions (sorry if my question makes that impression and causes the defensive reaction), but rather to provide a point of view on how this specification can be read.

**Comment by ljharb** [comment]
To be clear, I'm not a maintainer here and I've made no decisions :-) but to me, what I described seems like a reasonable reason for the restriction.

**Comment by bromps** [comment]
I agree with both of you.  The description of SemVer is much broader than the ruleset that it sets out to define.  I think some updates are needed.

**Comment by AndreKR** [comment]
Occasionally I find it useful to auto-generate the PATCH version from the number of commits between root and the release, causing it to increase steadily. I just found out that technically that's violating semver, and I'm also curious what the rationale behind the "MUST reset to zero" rule might be.

**Comment by jwdonahue** [comment]
Perhaps nobody noticed, but there is a loop-hole to this reset issue, that is implicitly built into the spec.  The spec says your version generator should follow the reset pattern:

0.1.0
...
1.0.0
...
2.0.0

But you are under no obligation to maintain that historical pattern in your release feeds. You are free to remove any version for any reason, so you could in practice, follow the OP's suggested pattern without breaking consumers, but there's nothing in the SemVer spec that provides any of the suggested new semantics that would be useful to consumers.

**Comment by jwdonahue** [comment]
From the 3rd paragraph of the [Introduction](https://semver.org/#introduction):

> As a solution to this problem, we propose a simple set of rules and requirements that dictate how version numbers are assigned and incremented. These rules are based on but not necessarily limited to pre-existing widespread common practices in use in both closed and open-source software.

These conventions have their roots in common manufacturing practices of the mid 1900's, and those evolved from earlier practices that themselves evolved from the publishing industry. I remember when bills-of-materials for boards included PROM's burned from "golden PROM's" created at the desk of an EE/SD and hand carried down to the manufacturing floor. Due to paper or computer form design and internal company practices, version labels on those GPROM's often adhered to the P{#}-Rev{alpha|num} format. Later, pure software version numbers were usually just a counter that we had to remember to bump in a header file somewhere, until we figured out how to automate that process.

The pure software developers eventually rediscovered that there was a reason for the MajorVersion-Revision formats their predecessors used, and often resorted to using fixed or floating point numbers. Decades later, the Major.Minor.Patch format evolved from those. Resetting lower order fields to zero wasn't just a convention, it evolved from the fact that we had limited space to store values and we had to reset from 1.9 to 2.0 or we'd run out of storage or display space.

**Comment by jwdonahue** [comment]
The fact that other semantic schemes can be mapped onto SemVer compliant strings and version histories, does not imply that SemVer should be modified to better accommodate or directly define them, or how they should be mapped onto SemVer. The spec, as it stands now, reflects the consensus view of stake-holders who maintain packaging/dependency tooling in common use. They are motivated to protect the status quo, which many argue is a good thing when it comes to standards.

Most of the SemVer compliant tooling also provide non-SemVer specific features, so you are often free to implement any scheme you see fit, and can often shoe-horn those into SemVer dependent ecosystems. Your mileage may vary.

**Comment by dylanopen** [comment]
I'm 4 years late to this. But I just wanted to add some reasons as to why I think that *not* resetting the minor and patch versions is actually a good idea. I think @nolar is completely right.

Looking at the issue number for a very similar issue (#9) it's clear that it was realised very early on that not resetting the minor and patch versions could work.

So why do I think this is a good idea?

- Not resetting any of the version segments allows you to easily tell the number of versions that have been released. If you have a project which uses this 'continuous version numbering' system, you can simply add up the major, minor and patch versions to find out exactly how many releases there have been.
- It can also be used to tell how far along a project is in development. If you see something like `4.0.0`, it's not clear as to whether it's had 4 released or 400, as the evidence of progress is essentially thrown away when resetting the minor and patch. Of course, project maturity shouldn't be judged based on the number of releases. But it's still a good way of telling roughly how mature a package is.
- As a library maintainer, I wouldn't like to go from, say, version `2.17.38` to just `3.0.0`. Maybe it's just a psychological thing, but I think that a library would look better for users having `3.17.38` instead as it shows it's actually had more active development.
- It's easy to see exactly how many patches there have been, just look at the last number.
- It's also easy to see exactly how many minor versions there have been, look at the middle number.
- Most importantly, you can very easily see whether or not a library regularly introduces breaking change. If a library has been around for 5 years and has a major version of `4`, it's probably more stable than one with a major version of `82` as it less frequently introduces breaking changes. (Before you comment, I know, you can still see this with 'resetting' semver. But most library maintainers will release breaking changes in 'big bundles' - I believe because of this notion of 'resetting the progress'. Continuous Semver, a term I will coin 😄, mostly gets rid of this issue.

- Why does it get rid of this issue? The biggest issue that semver has at the moment is library devs not wanting to push a new major version. They bundle up loads of breaking changes and push these all in one major release, making it really difficult for library users to update (as so much has changed).
- I'm the developer of a Rust library called `Realms` (https://github.com/dylanopen/realms). I use this continuous semver approach, because it allows me to push a new major version whenever I make a change, so it's really easy for my users to update (just changing a few function calls, usually).
- If I instead reset the minor and patch, I'd be much more weary of making major releases. That's because resetting any sort of versioning makes it feel like your losing progress. This is silly of course, but it's a real thing and the best we can do is try to make releasing major versions more normalised. Breaking change is never good, but it's much better to slowly release it rather than one huge change which makes migration near impossible.

This is of course my opinion. I'm well aware that most of my points ultimately come down to "bigger version numbers = shows more progress = good", so you may well disagree. For me though, there haven't been any downsides to continuous semver so I think I will keep using it into the future :)

**Comment by dylanopen** [comment]
Oh, one more thing I forgot to mention. With continuous versioning, it makes releasing important patches actually make sense.

Let's say you have a security vulnerability that affects these 6 versions:

- 1.3.5
- 1.3.6
- 2.3.6
- 2.4.6
- 3.4.6
- 3.4.7

You can just release these versions:

- 1.3.8
- 2.4.8
- 3.4.8

This would be really awkward without continuous semver.

**Comment by ljharb** [comment]
Why would that be really awkward? You'd release v1.3.7, v2.4.7, and 3.4.8, and the CVE and everyone's tools would Just Work (as they do now)

**Comment by ljharb** [comment]
That clarity isn't useful for versions to convey. Basically the only thing a semver version is meant to convey is amount of breakage.

You always need to read the release notes/changelog to know what's changed in a release. For a security issue, the _CVE_ tells you they're all related.

**Comment by dylanopen** [comment]
@ljharb That's true, semver isn't meant to convey how patches link together. I just like my version numbers to give that sort of info.

Ultimately we could debate this forever: it's just my own unpopular opinion of how to do semver! There are positives and negatives to each I guess, and although I haven't found it yet, there probably is a reason as to why continuous semver isn't widely adopted. I just like it, and that's why I use it for my open source projects :)
